testctrl/proto/test_sessions.proto

```diff
@@ -0,0 +1,165 @@
+  // the name of the Java worker may be "gcr.io/grpc-testing/java_qps_worker".
+  string name = 1;  // [(google.api.field_behavior) = REQUIRED];
+
+  // The kind which assigns the responsibilities of this container.
+  Kind kind = 2;  // [(google.api.field_behavior) = OPTIONAL];
+
+  oneof version_identifier {
+    // The hash assigned to the image.
+    string sha256 = 2;
+
+    // A tag which was explicitly set on the image.
+    //
+    // For example, a developer may explicitly tag an image for repetitive
+    // testing and sharing. They may add a tag with a PR number or version.
+    //
+    // Note: the "latest" tag is automatically set by the registry, and there
+    // are no guarantees that it will point to a specific image. It should be
+    // avoided.
+    string tag = 3;
+  }
+
+  // The number of pods that should be created with this container. For example,
+  // specify 2 replicas on a client container to create 2 clients.
+  //
+  // This field will always be ignored on driver containers. It is ignored on
+  // server containers at present, but may be supported eventually.
+  int32 replicas = 4;
+}
+
+message StartTestSessionRequest {
+  // The test scenarios to run as a part of the test session.
+  repeated grpc.testing.Scenario scenarios = 1;  // [(google.api.field_behavior) = REQUIRED];
+
+  // The container used as the driver which runs the test scenarios.
+  //
+  // If unset, the "latest" tag is used with the standard driver.
+  Container driver_container = 2;  // [(google.api.field_behavior) = OPTIONAL];
+
+  // The containers used as clients.
+  repeated Container worker_containers = 3;  // [(google.api.field_behavior) = REQUIRED];
+
+  // An optional location where the results should be written.
+  oneof result_destination {
+    // The name of the BigQuery table where results should be written.
+    string bq_results_table = 4;  // [(google.api.field_behavior) = OPTIONAL];
+
+    // The fully qualified GCS URI where a results file should be written.
+    string gcs_results_file_uri = 5;  // [(google.api.field_behavior) = OPTIONAL];
+  }
+ }
+
+message TestSessionResults {
+  // The results for each scenario.
+  repeated grpc.testing.ScenarioResult scenario_results = 1;
+
+  // The amount of time that this session lived, including scheduling to
+  // to reporting results.
+  google.protobuf.Duration time_elapsed = 2;
+}
+
+enum TestSessionPhase {
+  // The test session is waiting for idle nodes to schedule pods.
+  QUEUED = 0;
+
+  // Kubernetes is creating and starting the test pods.
+  PROVISIONING = 1;
+
+  // The driver is waiting for the workers to warmup.
+  WARMING_UP = 2;
+
+  // The tests are running and metrics are being collected.
+  TESTING = 3;
+
+  // Results are being saved to persistent storage.
+  REPORTING = 4;
+}
+
+message TestSessionMetadata {
+  // If the session has not terminated, this field provides the current phase
+  // of the test session.
+  TestSessionPhase current_phase = 1;
```

Comment:
I don't think this design is ideal. Two issues:

1. Does not say when the phase shifted.
2. Does not include an `unknown` phase as recommended by protobuf styleguide.

```proto
message Phase {
    enum Kind {
      // The test session has not terminated, but the phase could not be determind.
      UNKNOWN = 0;
      
      // The test session is waiting for idle nodes to schedule pods.
      QUEUED = 1;

      // Kubernetes is creating and starting the test pods.
      PROVISIONING = 2;

      // The driver is waiting for the workers to warmup.
      WARMING_UP = 3;

      // The tests are running and metrics are being collected.
      TESTING = 4;

      // Results are being saved to persistent storage.
      REPORTING = 5;
    }
  
    // The type of phase.
    Kind kind = 1;
    
    // The time that the phase was noticed to change by the service.
    google.protobuf.Timestamp shift_time = 2;
  }

  Phase current_phase = 1;
```

This also permits a history to be added in the future:

```proto
repeated Phase phase_history = 5;
```
Thoughts?